clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,63 +348,560 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+                    "are not currently assigned. Requesting a metadata update now to resolve " +
+                    "topic names.", assignmentUnresolved.keySet());
+            metadata.requestUpdate(true);
+        }
+    }
+
+    /**
+     * Look for topic in the global metadata cache. If found, add it to the local cache and
+     * return it. If not found, look for it in the local metadata cache. Return empty if not
+     * found in any of the two.
+     */
+    private Optional<String> findTopicNameInGlobalOrLocalCache(Uuid topicId) {
+        String nameFromMetadataCache = metadata.topicNames().getOrDefault(topicId, null);
+        if (nameFromMetadataCache != null) {
+            // Add topic name to local cache, so it can be reused if included in a next target
+            // assignment if metadata cache not available.
+            assignedTopicNamesCache.put(topicId, nameFromMetadataCache);
+            return Optional.of(nameFromMetadataCache);
         } else {
-            transitionToFailed();
-            throw new IllegalStateException("Cannot set new target assignment because a " +
-                    "previous one pending to be reconciled already exists.");
+            // Topic ID was not found in metadata. Check if the topic name is in the local
+            // cache of topics currently assigned. This will avoid a metadata request in the
+            // case where the metadata cache may have been flushed right before the
+            // revocation of a previously assigned topic.
+            String nameFromSubscriptionCache = assignedTopicNamesCache.getOrDefault(topicId, null);
+            return Optional.ofNullable(nameFromSubscriptionCache);
         }
     }
 
     /**
-     * Returns true if the member has a target assignment being processed.
+     * Build set of TopicPartition for the partitions included in the heartbeat topicPartitions,
+     * and using the given topic name.
      */
-    private boolean hasPendingTargetAssignment() {
-        return targetAssignment.isPresent();
+    private SortedSet<TopicPartition> buildAssignedPartitionsWithTopicName(
+            List<Integer> topicPartitions,
+            String topicName) {
+        SortedSet<TopicPartition> assignedPartitions = new TreeSet<>(COMPARATOR);
+        topicPartitions.forEach(tp -> assignedPartitions.add(new TopicPartition(topicName, tp)));
+        return assignedPartitions;
+    }
+
+    /**
+     * Revoke partitions. This will:
+     * <ul>
+     *     <li>Trigger an async commit offsets request if auto-commit enabled.</li>
+     *     <li>Invoke the onPartitionsRevoked callback if the user has registered it.</li>
+     * </ul>
+     *
+     * This will wait on the commit request to finish before invoking the callback. If the commit
+     * request fails, this will proceed to invoke the user callbacks anyway,
+     * returning a future that will complete or fail depending on the callback execution only.
```

Comment:
Should we trigger both in parallel?

Reply:
Yes, makes sense to me. It could be bundled up along with the callbacks triggering (both, revocation and assign), so that we trigger them all in the same poll iteration, while still making sure that they are executed in the right order. Filed [KAFKA-15835](https://issues.apache.org/jira/browse/KAFKA-15835) and I will address that as a follow-up right after this.